Framework/src/KafkaPoller.cxx

```diff
@@ -0,0 +1,166 @@
+#include "QualityControl/KafkaPoller.h"
+
+#include "QualityControl/QcInfoLogger.h"
+#include "kafka/KafkaException.h"
+#include "kafka/Properties.h"
+#include "proto/events.pb.h"
+#include <chrono>
+#include <ranges>
+
+namespace o2::quality_control::core
+{
+
+namespace proto
+{
+
+bool isRunNumberSet(int runNumber)
+{
+  return runNumber != 0;
+}
+
+bool isEnvironmentIdSet(const std::string_view environmentID)
+{
+  // qc is default value for Activity object so we need to check against it as well
+  return !environmentID.empty() && environmentID != "qc";
+}
+
+struct Ev_RunEventPartial {
+  std::string_view state;
+  events::OpStatus transitionStatus;
+  std::string_view environmentID;
+  int runNumber;
+};
+
+bool operator==(const events::Ev_RunEvent& runEvent, const Ev_RunEventPartial& runEvenPartial)
+{
+  // TODO: Should we check whether the error is empty?
+  if (runEvent.state() != runEvenPartial.state ||
+      runEvent.transitionstatus() != runEvenPartial.transitionStatus) {
+    return false;
+  }
+
+  if (isRunNumberSet(runEvenPartial.runNumber) && runEvent.runnumber() != runEvenPartial.runNumber) {
+    return false;
+  }
+
+  if (isEnvironmentIdSet(runEvenPartial.environmentID) && runEvent.environmentid() != runEvenPartial.environmentID) {
+    return false;
+  }
+
+  return true;
+}
+
+auto recordToEvent(const kafka::Value& kafkaRecord) -> std::optional<events::Event>
+{
+  events::Event event;
+
+  if (!event.ParseFromArray(kafkaRecord.data(), kafkaRecord.size())) {
+    ILOG(Error, Ops) << "Received wrong or inconsistent data while parser Event from kafka proto" << ENDM;
+    return std::nullopt;
+  }
+
+  return event;
+}
+
+void fillActivityWithoutTimestamp(const events::Event& event, Activity& activity)
+{
+  if (event.has_runevent()) {
+    auto& runEvent = event.runevent();
+
+    if (!isRunNumberSet(activity.mId)) {
+      activity.mId = runEvent.runnumber();
+    }
+
+    if (!isEnvironmentIdSet(activity.mProvenance)) {
```

Comment:
I think you mistook provenance with partition ID (which is another, somewhat obsolete name for environment ID). Provenance says whether the QC objects were created in sync, async or monte carlo QC.